### PR TITLE
Feature: Api framework

### PR DESCRIPTION
**Pull request by k9ert**
This is a continuation of the PR #713 but moving the code to a non-master branch on my remote.
Will close the original PR. With this PR, Specter gets an API. The documentation is in the docs-folder.

The API is only activated if:
```
export SPECTER_API_ACTIVE=true
```
It's also active by default in Dev-Mode.

Apart from this new feature:
* The test-framework needed some refactoring. Better some bug-fixing. The biggest change is that the `bitcoin_regtest` fixture changed scope from `module` to `session`. This means that the regtest-node is only be spun up exactly once for the tests.
* Made an update of a bunch of dependecies
* One of the dependencies (`cryptography`) needed a newer OS-version but the used cypress-image was based on an old debian-image. So i've created a new base-image ` docker/cypress-base-ubuntu-focal/Dockerfile` which is ubuntu based and what is from now on the new base for ` docker/cypress-python/Dockerfile`
* The `user.password` is now a property which will automatically detect a plain-password and will do the necessary salted hashing
* The wallet-enpoint for creating PSBT's got a refactoring and now we have `psbt_creator.py` very similiar to the `wallet_imported.py`. There are people calling that a "command-pattern".
* The `wallet_importer.py` got some improvements as well
* The `wallet.py` has a change where we're rather throwing an error than returning 0-balances. We should have more courage to raise `SpecterErrors`
* I've changed the port-numbers of adhoc-bitcoind-nodes spun up in tests and i've wrapped them in try/raise to make sure that they will shutdown in any case of result of the tests (several tests).
* Finally there is also a small bugfix for this:
k9ert, [21.06.21 14:26]
```
[2021-06-18 14:36:39,975] INFO in wallet_manager: Wallet already in Specter, updating wallet: testwallet_ghost
[2021-06-18 14:36:40,107] ERROR in controller: Uncaught exception: 'dict object' has no attribute 'devices_signed'
[2021-06-18 14:36:40,108] ERROR in controller: Traceback (most recent call last):
  File "/tmp/cirrus-ci-build/.env/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/tmp/cirrus-ci-build/.env/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/tmp/cirrus-ci-build/.env/lib/python3.8/site-packages/flask_login/utils.py", line 269, in decorated_view
    return func(*args, **kwargs)
  File "/tmp/cirrus-ci-build/src/cryptoadvance/specter/server_endpoints/wallets.py", line 683, in send_new
    return render_template(
  File "/tmp/cirrus-ci-build/.env/lib/python3.8/site-packages/flask/templating.py", line 137, in render_template
    return _render(
  File "/tmp/cirrus-ci-build/.env/lib/python3.8/site-packages/flask/templating.py", line 120, in _render
    rv = template.render(context)
  File "/tmp/cirrus-ci-build/.env/lib/python3.8/site-packages/jinja2/environment.py", line 1304, in render
    self.environment.handle_exception()
  File "/tmp/cirrus-ci-build/.env/lib/python3.8/site-packages/jinja2/environment.py", line 925, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/tmp/cirrus-ci-build/src/cryptoadvance/specter/templates/wallet/send/sign/wallet_send_sign_psbt.jinja", line 2, in top-level template code
    {% set tab = 'send' %}
  File "/tmp/cirrus-ci-build/src/cryptoadvance/specter/templates/wallet/components/wallet_tab.jinja", line 1, in top-level template code
    {% extends "base.jinja" %}
  File "/tmp/cirrus-ci-build/src/cryptoadvance/specter/templates/base.jinja", line 62, in top-level template code
    {% block main %}
  File "/tmp/cirrus-ci-build/src/cryptoadvance/specter/templates/wallet/components/wallet_tab.jinja", line 12, in block 'main'
    {% block content %}
  File "/tmp/cirrus-ci-build/src/cryptoadvance/specter/templates/wallet/send/sign/wallet_send_sign_psbt.jinja", line 199, in block 'content'
    <button type="button" class="btn signing-column-btn" id="{{ device.alias }}_tx_sign_btn" {{ 'disabled style=background-color:#303c49;' if device.alias in psbt["devices_signed"] else '' }}>
jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'devices_signed'
[2021-06-18 14:36:40,149] INFO in wallet_manager: Finished updating wallet:  testwallet_ghost
[2021-06-18 14:36:40,149] INFO in wallet_manager: Wallet already in Specter, updating wallet:
```

### CODE REVIEW

File: src/cryptoadvance/specter/api/API.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
+## Specter API
```
Comment (stepansnigirev) [inline src/cryptoadvance/specter/api/API.md:1]: Looks like this file is duplicated in two locations:
`src/cryptoadvance/specter/api/API.md` and `docs/API.md`.
Do we need both?

File: src/cryptoadvance/specter/api/rest/livereadyprobes.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+
+@rest_resource
+class ResourceLiveness(BaseResource):
+    """/api/healthz/liveness"""
+
+    endpoints = ["/healthz/liveness"]
+
+    def get(self):
+        return "i am alive"
```
Comment (stepansnigirev) [inline src/cryptoadvance/specter/api/rest/livereadyprobes.py:22]: If we return jsons is all get requests maybe makes sense to return json here as well, for example
`{"status": "i am alive"}`

File: src/cryptoadvance/specter/api/rest/livereadyprobes.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    def get(self):
+        try:
+            app.specter.check()
+        except Exception as e:
+            logger.error(f"Readyness probe failed:{e}")
+            # Would be cool to have a timeout check here to act more sophisticated e.g. warn in the logs
+            # or something but looks complicated:
+            # https://sqlalchemy.narkive.com/U4m4aqf9/set-a-query-timeout-on-a-per-query-basis
+        return "i am ready"
```
Comment (stepansnigirev) [inline src/cryptoadvance/specter/api/rest/livereadyprobes.py:39]: Same as above, I suggest returning jsons everywhere

File: src/cryptoadvance/specter/process_controller/node_controller.py [hunk 4]
```diff
@@ -231,8 +229,13 @@ def testcoin_faucet(self, address, amount=20):
             default_address = default_rpc.getaddressinfo(default_address)[
                 "unconfidential"
             ]
-        if balance < amount:
+        while True:
+            btc_balance = default_rpc.getbalance()
+            if isinstance(btc_balance, dict):  # elementsd
```
Comment (stepansnigirev) [inline src/cryptoadvance/specter/process_controller/node_controller.py:234]: This assumption is wrong. Our LiquidRPC class returns value only in Bitcoin unless you provide `asset=None` keyword.
See https://github.com/cryptoadvance/specter-desktop/blob/master/src/cryptoadvance/specter/liquid/rpc.py#L36-L53

Better to use `is_liquid(chain)` or check implementation from node info - Bitcoin Core or Elements Core

Comment (k9ert) [reply src/cryptoadvance/specter/process_controller/node_controller.py:234]: I've wrote a small test which is currently failing. Maybe you can check:
https://cirrus-ci.com/task/6114508184748032?logs=test#L183

File: src/cryptoadvance/specter/api/rest/api.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            logger.error(se)
+            if str(se).startswith(f"Wallet {wallet_alias} does not exist!"):
+                return abort(403, message=f"Wallet {wallet_alias} does not exist")
+            logger.error(se)
+            return abort(500)
+
+        wallet.get_balance()
+        wallet.check_utxo()
+        wallet.check_unused()
+        logger.info("Muuh")
```
Comment (ben-kaufman) [inline src/cryptoadvance/specter/api/rest/api.py:49]: Delete?

File: src/cryptoadvance/specter/api/rest/api.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+        return_dict = {}
+        address_index = wallet.address_index
+        validate_merkle_proofs = app.specter.config.get("validate_merkle_proofs")
+        logger.info("Meeh")
```
Comment (ben-kaufman) [inline src/cryptoadvance/specter/api/rest/api.py:54]: Delete?

File: src/cryptoadvance/specter/api/rest/api.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+        # Flatten the list
+        flat_list = []
+        for element in tx_list:
+            for dic_item in element:
+                flat_list.append(dic_item)
+
+        # Check if scanning
+        scan = wallet.rescan_progress
+        logger.info("Meeh3")
```
Comment (ben-kaufman) [inline src/cryptoadvance/specter/api/rest/api.py:71]: Delete?

File: docs/api.md [hunk 8]
```diff
@@ -0,0 +1,606 @@
...
+
+# Endpoints
+
+
+## API Endpoint [/api/healthz/liveness](/api/healthz/liveness) (GET)
+This endpoint works as heathz-check. See e.g. here:
+https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/
+
+## API Endpoint [/api/healthz/readyness](/api/healthz/readyness) (GET)
+This endpoint works as heathz-check. See e.g. here:
```
Comment (ben-kaufman) [inline docs/api.md:33]: typo heathz -> health ?

Comment (k9ert) [reply docs/api.md:33]: It's indeed a "standard".

File: src/cryptoadvance/specter/internal_node.py [hunk 9]
```diff
@@ -119,29 +120,48 @@ def json(self):
...
-            self._bitcoind.stop_bitcoind()
-        self.bitcoin_pid = False
+        logger.info(f"STOPPING bitcoind {self.name} from status {self.status}")
+        success = self._bitcoind.stop_bitcoind()
+        if success:
+            self._update_status(None, self.DOWN)
+            logger.info(f"bitcoind {self.name} stopped")
+        else:
+            logger.error(f"Failed to stop bitcoind {self.name}")
+            self.status = self.BROKEN
```
Comment (ben-kaufman) [inline src/cryptoadvance/specter/internal_node.py:163]:
```suggestion
            self._status = self.BROKEN
```

**Pull request merged**